src/main/java/com/exasol/glue/filesystem/S3FileSystem.java

```diff
@@ -0,0 +1,140 @@
+    }
+
+    /**
+     * Creates a new instance of {@link S3FileSystem}.
+     *
+     * @param s3Client s3 client object
+     */
+    public S3FileSystem(final S3Client s3Client) {
+        this.s3Client = s3Client;
+    }
+
+    /**
+     * Checks if a given bucket exists.
+     *
+     * @param bucketName name of a bucket
+     * @return {@code true} if path is a directory, {@code false} otherwise
+     */
+    public boolean doesBucketExist(final String bucketName) {
+        try {
+            s3Client.headBucket(HeadBucketRequest.builder().bucket(bucketName).build());
+            return true;
+        } catch (final NoSuchBucketException exception) {
+            return false;
+        }
+    }
+
+    /**
+     * Deletes a given bucket contents.
+     *
+     * @param bucketName name of a bucket
+     */
+    public void deleteBucket(final String bucketName) {
+        LOGGER.info(() -> "Deleting S3 bucket '" + bucketName + "'.");
+        deleteObjects(bucketName, Optional.empty());
+    }
+
+    /**
+     * Deletes a given bucket with bucket key contents.
+     *
+     * @param bucketName name of a bucket
+     * @param bucketKey  bucket key value
+     */
+    public void deleteBucketKey(final String bucketName, final String bucketKey) {
+        LOGGER.info(() -> "Deleting S3 bucket '" + bucketName + "' with bucket key '" + bucketKey + "'.");
+        deleteObjects(bucketName, Optional.of(bucketKey));
+    }
+
+    private void deleteObjects(final String bucketName, final Optional<String> bucketKey) {
+        try {
+            final List<S3Object> objects = listObjects(bucketName, bucketKey);
+            List<ObjectIdentifier> objectIdentifiers = objects.stream() //
+                    .map(object -> ObjectIdentifier.builder().key(object.key()).build()) //
+                    .collect(Collectors.toList());
+            deleteObjectIdentifiers(bucketName, objectIdentifiers);
+        } catch (final SdkClientException exception) {
+            throw new ExasolConnectionException(
+                    ExaError.messageBuilder("E-EGC-19")
+                            .message("Failed to delete objects in {{BUCKET}} with key {{KEY}}.", bucketName,
+                                    bucketKey.orElse("emptyBucketKey"))
+                            .mitigation("Please check that credentials and bucket name are correct.").toString(),
+                    exception);
+        } catch (final S3Exception exception) {
+            throw new ExasolConnectionException(ExaError.messageBuilder("E-EGC-20")
+                    .message("Failed to delete objects in {{BUCKET}} with key {{KEY}} because of unknown S3 exception.")
+                    .parameter("BUCKET", bucketName).parameter("KEY", bucketKey.orElse("emptyBucketKey"))
+                    .ticketMitigation().toString(), exception);
+        }
+    }
+
+    private List<S3Object> listObjects(final String bucketName, final Optional<String> bucketKey) {
```

Comment:
that method would run out of memory for buckets with lots of files.

Reply:
Same as above.